### PR TITLE
Typewriter experience

### PR DESCRIPTION
**Pull request by ellatrix**
## Description

This change maintains the caret (insertion point) position while the user is typing.

Text wrapping and `Enter`.

typewriter

iOS (there's still movement, but MUCH better). Apologies, Simulator doesn't correctly set the correct keyboard layout...

m-type

## How has this been tested?

Make sure there is enough content so that the page is scrollable.

On various devices test:

* `Enter`
* `Shift+Enter`
* `Backspace`
* Text wrapping (when text wraps to a new line, the position should update).

## Checklist:
- [x] My code is tested.
- [x] My code follows the WordPress code style. 
- [x] My code follows the accessibility standards. 
- [x] My code has proper inline documentation. 
- [x] I've included developer documentation if appropriate.

### CODE REVIEW

**Comment by jasmussen** [comment]
I like this **a lot**. 

I like that:

- The iOS editing experience is VASTLY improved (going from broken to usable) in this PR. 
- I like that the solution works the same across mobile and web.
- I like that the Enter behavior works regardless of where you are in the document. 

Here's a GIF of the iOS 12 simulator with this:

iphone

Even though the above is a little jumpy, it is VASTLY better than what is in master now, which is nigh unusable. Thanks so much for working on this, mobile is critical path.

I am in the process of installing the beta version of Xcode so I can test in iOS 13. Some changes have been made to how scrolling containers work there, and I want to test whether the jumpiness is the same or not in that. I'll return with more info. But for now, 👍 👍 👍 👍

File: packages/block-editor/src/components/block-list/moving-animation.js
Comment (youknowriad) [inline packages/block-editor/src/components/block-list/moving-animation.js]: Ideally, the animation stays agnostic to the store. At some point in the future we might want to move it to a generic hook. We can use the `enableAnimation` argument to pass this as an argument from the outside.

Comment (ellatrix) [reply packages/block-editor/src/components/block-list/moving-animation.js]: @youknowriad That was my initial thought as well, but there's a comment at the bottom of the parent component saying that it should be limited to checking typing for the selected block to reduce re-renders. We could wrap `useMovingAnimation` when the time comes to reuse it elsewhere?

**Comment by ellatrix** [comment]
@jasmussen Great! I tried a lot of different thing today to try to get rid of the jumpiness on iOS (which was there previously as well, and also, nowhere else do I see this problem), but unfortunately I find nothing that fixes it. Something I tried was scroll-locking the container at key and releasing it at keyup, but that doesn't change anything. It would also not be reliable to do that; the key down event can be cancelled resulting it to be locked forever. I'd appreciate other ideas for the iOS issue. Anyway, this is already much better. :)

**Comment by jasmussen** [comment]
I strongly agree.

I think the jumpiness comes from the soft keyboard imperceptibly quickly opening and closing. The soft keyboard on iOS is notoriously hard to work with. When the caret is in text, it stays open, when something that isn't text has focus, it closes. My theory is that because the caret jumps from one text block to a new empty richtext field, in the transition there is a brief period where focus isn't in text, starting the soft keyboard closing animation.

**Comment by ellatrix** [comment]
@jasmussen Ah, that's a good point... I wonder if we can somehow trick it into thinking it stays in a text field. :) Probably not.

**Comment by ellatrix** [comment]
@jasmussen You're right, when you use the arrow icons to tab through the blocks, I observe the same jumpiness.

**Comment by ellatrix** [comment]
Changed it now so that the scroll adjustment a lot smoother. With that I mean for everything except iOS. Before you might have sometimes seen the screen flicker. This should now be gone. :)

**Comment by ellatrix** [comment]
Strange that the tests I created are again successful locally but fail on Travis...

**Comment by ellatrix** [comment]
The only thing left here is figuring out while the newly added tests are failing with Travis (and pass locally).

**Review by jorgefilipecosta — COMMENTED** [review]
I added a commit that removes the usage useSelect inside useMovingAnimation hook.
In my tests, I found a small regression.
If we add many blocks (more than the visible area), and then we multiple select a large chunk of blocks and delete them by pressing backspace, the last blocks available blocks become invisible.

Jul-18-2019 13-12-25
(after backspace the last visible block should be the 60)

File: packages/block-editor/src/components/block-list/block.js
Comment (youknowriad) [inline packages/block-editor/src/components/block-list/block.js]: This is buggy until we merge #16669

Comment (ellatrix) [reply packages/block-editor/src/components/block-list/block.js]: So is this blocked by #16669?

Comment (youknowriad) [reply packages/block-editor/src/components/block-list/block.js]: This could be rewritten to use an object (if we want to merge before that PR)

**Comment by ellatrix** [comment]
@jorgefilipecosta The multi-select issue should be resolved. The last thing left here seems to be getting the new e2e tests to work on Travis.

File: packages/block-editor/src/components/block-list/block.js
Comment (ellatrix) [inline packages/block-editor/src/components/block-list/block.js]: Somehow, this making all tests in `packages/e2e-tests/specs/navigable-toolbar.test.js` fail.

**Comment by ellatrix** [comment]
E2e tests pass 🎉
This is ready for a final review.

File: packages/block-editor/src/components/writing-flow/index.js
Comment (aduth) [inline packages/block-editor/src/components/writing-flow/index.js]: Is there a specific reason the logic needs to occur in this component? It's already quite large, and is primarily focused around interactions _between_ blocks.

To me, the proposed behavior reminds me a fair bit of the [`<MultiSelectScrollIntoView>` component](https://github.com/WordPress/gutenberg/tree/master/packages/block-editor/src/components/multi-select-scroll-into-view), which can be used (or not used) independently.

Comment (ellatrix) [reply packages/block-editor/src/components/writing-flow/index.js]: Yes, that would be good!

Comment (ellatrix) [reply packages/block-editor/src/components/writing-flow/index.js]: Done

File: packages/block-editor/src/components/writing-flow/index.js
Comment (aduth) [inline packages/block-editor/src/components/writing-flow/index.js]: What's the reason for this change?

I seem to recall some need for this to imitate `BlockListBlock`:

https://github.com/WordPress/gutenberg/blob/916cbb0817f3c60b54067b361f56475acf511cee/packages/block-editor/src/components/block-list/block.js#L395-L398

File: packages/block-editor/src/components/typewriter/index.js
Comment (aduth) [inline packages/block-editor/src/components/typewriter/index.js]: `compose` isn't strictly necessary here being a single higher-order component.

```suggestion
export default withSelect( ( select ) => {
	const { getSelectedBlockClientId } = select( 'core/block-editor' );
	return { selectedBlockClientId: getSelectedBlockClientId() };
} )( TypeWriter );
```

File: packages/block-editor/src/components/typewriter/index.js
Comment (aduth) [inline packages/block-editor/src/components/typewriter/index.js]: As a new component, I might have been curious to see if it could have been implemented as a function component using hooks (could simplify especially event binding with `useEffect` and handling with `useCallback`, cancellations on unmount, etc).

https://reactjs.org/docs/hooks-reference.html#cleaning-up-an-effect

File: packages/edit-post/src/components/visual-editor/style.scss
Comment (aduth) [inline packages/edit-post/src/components/visual-editor/style.scss]: I suppose it's intended then that there's a large gap at the bottom of a scrollable post?

image

File: packages/block-editor/src/components/typewriter/index.js
Comment (aduth) [inline packages/block-editor/src/components/typewriter/index.js]: Technically the attribute value could be `'false'`, which (ironically) is truthy.

https://developer.mozilla.org/en-US/docs/Web/HTML/Global_attributes#attr-contenteditable

Perhaps instead, use the `isContentEditable` boolean DOM property:

```suggestion
			document.activeElement.isContentEditable
```

https://developer.mozilla.org/en-US/docs/Web/API/HTMLElement/isContentEditable

File: packages/block-editor/src/components/typewriter/index.js
Comment (aduth) [inline packages/block-editor/src/components/typewriter/index.js]: Where is this return value used?

**Edit:** Okay, after thinking on it, I guess you're relying on `computeCaretRect` as a side-effect, and using `return` as a conciseness convenience. I'll let my initial confusion stand on its own as an anecdote in favor of the explicit:

```suggestion
			this.computeCaretRect();
			return;
```

Further, maybe `computeCaretRect` could be more explicit that it has the side-effect in being responsible to _assign_ a value, e.g. `assignCaretRect`.

File: packages/block-editor/src/components/typewriter/index.js
Comment (aduth) [reply packages/block-editor/src/components/typewriter/index.js]: Aside: If you choose to implement this as a function component, we can consider `useSelect` hook instead.

https://github.com/WordPress/gutenberg/tree/master/packages/data#useSelect

File: packages/block-editor/src/components/typewriter/index.js
Comment (ellatrix) [reply packages/block-editor/src/components/typewriter/index.js]: I think either could be fine. I wanted to check whether the element has the attribute, while `isContentEditable` is defined for any element. So a child of an editable field is also editable, but wouldn't have the `contenteditable` attribute. In this case either works fine I guess :)

File: packages/edit-post/src/components/visual-editor/style.scss
Comment (ellatrix) [reply packages/edit-post/src/components/visual-editor/style.scss]: Yes, that's by intention

File: packages/block-editor/src/components/writing-flow/index.js
Comment (ellatrix) [reply packages/block-editor/src/components/writing-flow/index.js]: Not sure if it makes much difference. Without the class, clicks anywhere under the block list will redirect. With the class, only clicks vertically under it.

**Comment by aduth** [comment]
After having written a post while on this branch, one thing I'm not so sure about is whether <kbd>ArrowUp</kbd> and <kbd>ArrowDown</kbd> should keep the caret positioned where it was placed. I can see where this aligns with the goal of the pull request, but from a writing perspective, it feels unnatural and unexpected.

I don't think it's a blocker, but it could be worth revisiting pending feedback about the feature.

On that note, is there any idea to make this opt-in or opt-out? I could see how it might not be for everyone, since it doesn't behave like most other text editors.

**Comment by aduth** [comment]
I have a hard time reproducing it, but multiple times now I've experienced when simply trying to click in a paragraph will result in the page scrolling and an unintended multi-selection.

**Comment by ellatrix** [comment]
@aduth Thanks for testing! Yes, the arrow keys should perhaps be excluded, but I kind of liked the behaviour.

> I have a hard time reproducing it reliably, but multiple times now I've experienced when simply trying to click in a paragraph will result in the page scrolling and an unintended multi-selection.

I assume this is specific to this branch? I'll have a look and try to reproduce.

**Comment by aduth** [comment]
> > I have a hard time reproducing it reliably, but multiple times now I've experienced when simply trying to click in a paragraph will result in the page scrolling and an unintended multi-selection.
> 
> I assume this is specific to this branch? I'll have a look and try to reproduce.

I think so, yes, though I haven't done any extensive writing on the `master` branch lately. I assumed it was related to the scroll behaviors introduced here (if scroll occurs during the mousedown and could cause the multi-selection to be triggered).

**Comment by draganescu** [comment]
This is great! Amazing work @ellatrix :)

I think @aduth's issue happens when you leave a tab for a while open and then come back to it. There may be some suspension done by the browser that messes with our JS.

File: packages/block-editor/src/components/typewriter/index.js
Comment (ellatrix) [reply packages/block-editor/src/components/typewriter/index.js]: I'm not sure how you'd convert the properties that I'm setting. E.g. `this.caretRect`. Using state here adds extra rerenders.

**Comment by jasmussen** [comment]
> On that note, is there any idea to make this opt-in or opt-out? I could see how it might not be for everyone, since it doesn't behave like most other text editors.

Off the bat it's important to note that this absolutely has to be always-on at the very least on any iOS device, but perhaps for simplicity just anything mobile. The iOS experience is utterly broken without it. 

For the rest, I personally super duper like the typewriter effect, and although this is a heirloom locally sourced anecdotal datapoint of one, I feel people will like it too. Except perhaps the arrow-key typewrite effect, while I like it I'm not as convinced on that one for the desktop. But the short of it is, if we are to make a desktop opt-out, we'd want to think long and hard about the default first, though. I'm obviously in the "on by default" category.

**Comment by ellatrix** [comment]
I'm also all for having this experience be the default experience. I do really like the arrow key behaviour, but maybe I'm alone in wanting that on by default.

But you all know how a typewriter works right? If you turn the know, the page "scrolls" up and down. 😉

What else is left here? If we'd like to use hooks for this new component, I might need some help from someone who is very good with them. I've been thinking about how to do it, but I'm not sure how I can get rid of the class property.

**Comment by draganescu** [comment]
This works amazing :)

File: packages/block-editor/README.md
Comment (draganescu) [inline packages/block-editor/README.md]: Is this the same vertical distance from the top or bottom edges of the viewport?

File: packages/block-editor/src/components/typewriter/index.js
Comment (draganescu) [reply packages/block-editor/src/components/typewriter/index.js]: Is this something blocking for this feature? I ask because it feels as a great UX improvement and it could always be refactored later in a separare issue/PR, right?

File: packages/block-editor/README.md
Comment (ellatrix) [reply packages/block-editor/README.md]: Both? :)

File: packages/block-editor/src/components/typewriter/index.js
Comment (ellatrix) [reply packages/block-editor/src/components/typewriter/index.js]: I don't think this is blocking, but something nice to have. I'm just unsure how to do it properly. Might need some help there :D

File: packages/block-editor/src/components/writing-flow/index.js
Comment (draganescu) [reply packages/block-editor/src/components/writing-flow/index.js]: Doesn't the old behaviour make more sense?

File: packages/block-editor/README.md
Comment (draganescu) [reply packages/block-editor/README.md]: Yes but it feels as if this depends on where my caret is ... In other editors typewriter mode made it always have the current line at 1/2 the viewport size. Here, although I like it, it is varying. What I meant by my comment is to somehow describe this behaviour here with a longer phrase.

File: packages/edit-post/src/components/visual-editor/style.scss
Comment (draganescu) [inline packages/edit-post/src/components/visual-editor/style.scss]: @jasmussen had a comment on this other simpler PR about using VH https://github.com/WordPress/gutenberg/pull/15165#issuecomment-486573474 as a unit having that varying bottom space is inconsistent as different resolutions. Not sure it applies here though.

Comment (jasmussen) [reply packages/edit-post/src/components/visual-editor/style.scss]: It partially applies only insofar as we should use the vh unit only when we really want to.

In that other PR, the VH unit added margin to the bottom of the editor, sized in vh. Which meant the margin was different depending on viewport height.

In this PR, and correct me if I'm wrong, the vh unit intends to keep _the line of text you are writing_ at the center of the screen. If that is correct, then it seems to make sense given the typewriter writing experience.

File: packages/block-editor/src/components/block-list/index.js
Comment (youknowriad) [inline packages/block-editor/src/components/block-list/index.js]: I personally like seeing the animation while hitten "Enter" or using the slash inserter. Let's have more opinions here. @jasmussen @kjellr @mapk @mtias

File: packages/block-editor/src/components/typewriter/index.js
Comment (youknowriad) [reply packages/block-editor/src/components/typewriter/index.js]: Properties are easily replaced with `useRef`

**Review by youknowriad — COMMENTED** [review]
Am I the only one finding it a bit strange? Especially the fact that it starts scrolling while there's a lot of room in my screen still. I guess the idea is to keep the end of the content in the middle of the screen, Maybe it's better if it's more down the page. 

Also, should we consider adding an option to enable/disable this behavior? I'm not sure it's something that benefits everyone?

**Comment by senadir** [comment]
>Also, should we consider adding an option to enable/disable this behavior? I'm not sure it's something that benefits everyone?

I think this is a good thing to have yeah, it's not always wanted, especially when the content is short or you just started working on it.
maybe an opt in should be a good start

**Comment by jasmussen** [comment]
> Am I the only one finding it a bit strange? Especially the fact that it starts scrolling while there's a lot of room in my screen still. I guess the idea is to keep the end of the content in the middle of the screen, Maybe it's better if it's more down the page.

There are a couple of bits to this. So as to unstick the various bits, I'm going to GIF each piece of the puzzle, so the full picture is more easily spotted. Basically there are two ingredients. 

## 1. The typewriter writing experience

 basic typewriter

This behavior is **identical to what is running on medium.com**. The only difference is **when it kicks in**. In this PR, the typewriter effect kicks in at about the middle of the screen, at `50vh`. On Medium it kicks in at around the end:

medium

## 2. The arrowkey typewriter behavior**

arrowkeys

This feature is an addition to the _writing_ effect. It keeps the caret centered even when you're using the arrow keys to navigate. 

**It is important to note that both of the above are critical bugfixes for the iOS experience. We have to ship those, both of them, at the very least for iOS. As far as I'm concerned, this is not optional, it is a choice between supporting the iPhone, or not.**

## Next Steps

I personally find that both of these effects as they are implemented make for a much improved writing experience. However the two bits that I imagine make people stumble are the **forced arrowkey typewriter effect**, and the fact that the **typewriter effect starts at the middle of the screen, rather than near the bottom**. 

I believe if we tune those two features, I have a hard time imagining why anyone would opt out. Perhaps to help jog peoples memories, this is what the writing flow is like, when you're writing longer articles:

current shipping

👆Please watch that GIF to the end, the writing experience for longer post is really not that good. And this PR makes it vastly better, and I'm surprised there isn't more exuberant excitement about this.

So here's what I think we should do:

1. Ship this for iPhone as fast as we can. Again, I don't see this as optional, but a critical bugfix that fixes an unusable shipping experience.

2. Tweak the _writing part_ of the typewriter effect to kick in not at the middle of the screen, but at maybe 75 or 80% of the screen height. So we have about this much space below the last paragraph:

Screenshot 2019-08-14 at 09 06 16

3. For now, toggle off the arrow-key navigation in code, and keep it there until we can ship this PR and guage the reaction. Then we can either try turning that on separately, or make it an option.

**Comment by tellthemachines** [comment]
I hate to be the party pooper but these changes break IE11 completely: every keystroke makes the page jump back to the top, making it impossible to write. If we're still supporting IE, we might want to look at disabling this functionality for that browser.

**Comment by mapk** [comment]
> and I'm surprised there isn't more exuberant excitement about this.

🎉 🌟 ❣️ 😃 

I'm in love with this feature! If we can include @jasmussen's [points above](https://github.com/WordPress/gutenberg/pull/16460#issuecomment-521128687), that would be a big win for this typewriting experience. 

> - Tweak the writing part of the typewriter effect to kick in not at the middle of the screen, but at maybe 75 or 80% of the screen height. So we have about this much space below the last paragraph.
> - For now, toggle off the arrow-key navigation in code, and keep it there until we can ship this PR and guage the reaction. Then we can either try turning that on separately, or make it an option.

Also...

> I hate to be the party pooper but these changes break IE11 completely

I'm open to disabling it for IE unless a fix is rather simple.

**Comment by draganescu** [comment]
🎉 🌟 ❣️ 😃 🎉 🌟 ❣️ 😃

**Comment by ellatrix** [comment]
Thanks for all the feedback, everyone!

> 2\. Tweak the _writing part_ of the typewriter effect to kick in not at the middle of the screen, but at maybe 75 or 80% of the screen height. So we have about this much space below the last paragraph:

I disagree with this. :) One of the reasons behind the typewriter experience is allowing the user to write at the middle of the screen instead of forcing it at the bottom of the screen. Without the extra space of **at least** `50vh`, this is not possible, unless you don't write at the end of the page, or you add a bunch of paragraphs at the end of the post so you can scroll back a bit. Cc @mtias and @shaunandrews I believe.

I want to emphasise that you are **not** forced to write at the middle of the screen. You can scroll the caret at any position and it will stay there, so it's up to you if you like writing at the middle or closer to the bottom.

If it were up to me I would even make `90vh` or something around that value, so you can write near the top of the page as well. This is one of the things I like about VSCode: there's some space that's 100% the height of the scroll container, minus one line.

Screenshot 2019-08-19 at 12 30 31

So, depending on what I want to do (let's say, write a new section), I can scroll the caret to the top and have some space.

If I need some context (literally), I can scroll the caret more down to see the text above, and I could write more towards the bottom of the page.

What I do get is that, when you first start writing a page, you don't **intentionally** set the caret at the middle, it just happens to put itself there, as you write more. I agree that triggering the typewriter effect for many people may not be ideal. They didn't intentionally scroll to that point. They're just writing, and suddenly the typewriter effect kicks in. I believe this is the problem you'd like to solve, @jasmussen?

I just don't think it should be solved by reducing the empty space on the bottom, but by setting some extra criteria for starting the typewriter effect. I'll have a look into this as soon as possible.

> For now, toggle off the arrow-key navigation in code, and keep it there until we can ship this PR and guage the reaction. Then we can either try turning that on separately, or make it an option.

I'm not sure if we'll ever do it then. I wonder if we cannot leave it on at the start and see how people find it? Then turn it off if it's too much?

> I'm open to disabling it for IE unless a fix is rather simple.

Let's just disable for IE. :)

> Setting the caret on a paragraph, scrolling until that paragraph is outside the viewport and then typing will not bring that paragraph back into view. On current master when repeating those steps the paragraph will be scrolled back into view.

That's an interesting scenario. :) I'll try to find a workaround.

> Adding a Media & Text block close to the bottom of the page and then typing in the text area often causes a jump where that block is repositioned in the centre of the page. I haven't been able to reproduce this one every single time, but it happens fairly often with this block. Not with any other I've tested so far.

Thanks, I'll do some testing.

**Comment by jasmussen** [comment]
> What I do get is that, when you first start writing a page, you don't intentionally set the caret at the middle, it just happens to put itself there, as you write more. I agree that triggering the typewriter effect for many people may not be ideal. They didn't intentionally scroll to that point. They're just writing, and suddenly the typewriter effect kicks in. I believe this is the problem you'd like to solve, @jasmussen?

Wow, you are completely right. Yes, this is exactly the point. Here's a GIF to help illustrate:

typewriter

My understanding was that the typewriter effect would always keep you at the center of the screen. No, as Ella kindly explained, it keeps you where you set the caret. 

If you are typing from the top, though, this also does mean that it will _kick in_ at the middle of the screen, and yes, that's what puts me off a little bit. I can't help but feel like this is a little _too smart_. Kicking in _later_ the first time you write, if this is easy to do, would probably address this. But I'd also be willing to this as is, and/or defer to others.

> I'm not sure if we'll ever do it then. I wonder if we cannot leave it on at the start and see how people find it? Then turn it off if it's too much?

Another thing I'm personally fine to test, especially since the first target is the plugin. But this is the one thing about this PR that makes it feel _not_ like an editor. When I scan with my eyes I lose the place because everything scrolls always, whereas when the paragraph sits still most of the time it's easier to scan at a glance. 

> Let's just disable for IE. :)

+1+1

File: packages/block-editor/src/components/block-list/index.js
Comment (ellatrix) [reply packages/block-editor/src/components/block-list/index.js]: I personally don't like it on `Enter` :D It really distracts to writing and you kind of expect it to be instant.

**Comment by ellatrix** [comment]
Ok, thanks! I will:

* Stop the typewriter effect from kicking in when the caret moves to fill up the page.
* Disable arrow keys for now.
* Disable for IE.

**Comment by ellatrix** [comment]
@jasmussen Does the latest commit 36ee38f fix the issue?

**Comment by jasmussen** [comment]
>  Does the latest commit 36ee38f fix the issue?

Thanks so very much for looking at this stuff. Here's what I'm seeing:

typewriter v2

That is — the page has to kind of "page down" one screen at initial writing, to kick in. Which feels like an improvement. And to be clear, this PR even in its current state feels like a vastly better writing experience than master.

I'm not sure how feasible it is, but I was picturing it kicking in around the space in the GIF where I type "here", and then stay there with no page jump.

**Comment by ellatrix** [comment]
> That is — the page has to kind of "page down" one screen at initial writing, to kick in. Which feels like an improvement. And to be clear, this PR even in its current state feels like a vastly better writing experience than master.
> 
> I'm not sure how feasible it is, but I was picturing it kicking in around the space in the GIF where I type "here", and then stay there with no page jump.

The "page jump" you're describing is the default browser behaviour. When you reach the end of the page, the browser will scroll the page. I think it's possible to do what you're describing... let me try. :)

**Comment by ellatrix** [comment]
@tellthemachines Could you re-test the problems you were experiencing. They should be fixed.

**Comment by ellatrix** [comment]
@jasmussen I fixed the page jump. But it will still only kick in when you reach the end. You'd have to manually scroll to a position that you'd like to stay in. I'll see if I can let it kick in around 75%.

**Comment by ellatrix** [comment]
@jasmussen Made it now kick when 70% of the writing surface is filled.

File: packages/block-editor/src/components/typewriter/index.js
Comment (ellatrix) [reply packages/block-editor/src/components/typewriter/index.js]: How would that work?

**Comment by jasmussen** [comment]
😍

My golly. In my opinion this is completely delightful. Here's what I see:

typewriter v3

This feels so good. I would welcome a sanity check by others testing, but I find this just delightful.

File: packages/block-editor/src/components/typewriter/index.js
Comment (tellthemachines) [inline packages/block-editor/src/components/typewriter/index.js]: We need to return `this.props.children` here instead of just `null`, otherwise nothing will render inside the editor content area.

**Review by tellthemachines — DISMISSED** [review]
The Chrome issues I'd mentioned are all fixed (thanks!) but IE now shows no content at all in the editor. I left a comment in the code regarding that. Other than that, it looks amazing 🎉

File: packages/block-editor/src/components/typewriter/index.js
Comment (ellatrix) [reply packages/block-editor/src/components/typewriter/index.js]: Whoops. :)

**Review by youknowriad — COMMENTED** [review]
This is looking way better. Two last remarks from me.

 - When used with metaboxes, there's a lot of empty space, I wonder if we should disable this if we have metaboxes.
 - I still feel like we may want to keep the animation.

**Comment by ellatrix** [comment]
> When used with metaboxes, there's a lot of empty space, I wonder if we should disable this if we have metaboxes.

I can remove just the space when there are meta boxes. Looks like the meta boxes themselves will give enough space.

> I still feel like we may want to keep the animation.

Other opinions on the animation? I just don't like it at all on typing. 🤷‍♀️

**Comment by jasmussen** [comment]
I tend to agree with no animation on typing, simply because I can't recall it happening in any other editor I've used. However I think it would be fine to test this out in a separate PR, so we can sort of compare the two side by side.

**Comment by youknowriad** [comment]
In some occasions, I scroll and start typing, the scroll position reverts to the previous one (before scrolling). It doesn't happen consistently though, I reproduce it 50% of the time.?

**Comment by ellatrix** [comment]
> In some occasions, I scroll and start typing, the scroll position reverts to the previous one (before scrolling). It doesn't happen consistently though, I reproduce it 50% of the time.?

Could you make a GIF or video?

**Comment by youknowriad** [comment]
scroll type

Notice in the gif above, it doesn't happen consistently. (The jumps)

**Comment by ellatrix** [comment]
@youknowriad Is that in Chrome or Firefox?

**Comment by youknowriad** [comment]
It's Chrome

**Comment by ellatrix** [comment]
@youknowriad It looks like it's caused by the 100ms debounce. The caret position will not have updated during scrolling (and continued smooth scrolling), and not until 100ms after it ends. It seems that this is debouncing is just not the way to go here. I'll rewrite this part with continuous animation frame requests.

**Review by youknowriad — APPROVED** [review]
LGTM 👍 

Did you try running the performance tests and compare with master? I don't really expect a change but you never know.

File: packages/block-editor/src/components/typewriter/index.js
Comment (youknowriad) [reply packages/block-editor/src/components/typewriter/index.js]:
```
const  myProperty = useRef( 'initial value' );

// Using the value
console.log( myProperty.current );

// Assigning a new value
myProperty.current =  'some value';
```

**Comment by ellatrix** [comment]
There's no noticeable difference between `master` and this branch.

```
Ellas-MacBook-Pro:gutenberg ella$ git checkout master
Switched to branch 'master'
Your branch is up to date with 'origin/master'.
Ellas-MacBook-Pro:gutenberg ella$ npm run test-performance

> gutenberg@6.3.0 test-performance /Users/ella/gutenberg
> wp-scripts test-e2e --config packages/e2e-tests/jest.performance.config.js

 PASS  packages/e2e-tests/specs/performance.test.js (105.638s)
  Performance
    ✓ 1000 paragraphs (101351ms)

Test Suites: 1 passed, 1 total
Tests:       1 passed, 1 total
Snapshots:   0 total
Time:        106.125s
Ran all test suites.

Average time to load: 4721ms
Average time to DOM content load: 4432ms
Average time to type character: 53.99ms
Slowest time to type character: 128ms
Fastest time to type character: 47ms
		
Ellas-MacBook-Pro:gutenberg ella$ git checkout -
Switched to branch 'try/typewriter'
Your branch is up to date with 'origin/try/typewriter'.
Ellas-MacBook-Pro:gutenberg ella$ npm run test-performance

> gutenberg@6.3.0 test-performance /Users/ella/gutenberg
> wp-scripts test-e2e --config packages/e2e-tests/jest.performance.config.js

 PASS  packages/e2e-tests/specs/performance.test.js (104.662s)
  Performance
    ✓ 1000 paragraphs (101565ms)

Test Suites: 1 passed, 1 total
Tests:       1 passed, 1 total
Snapshots:   0 total
Time:        104.739s, estimated 106s
Ran all test suites.

Average time to load: 4331ms
Average time to DOM content load: 4028ms
Average time to type character: 53.84ms
Slowest time to type character: 133ms
Fastest time to type character: 48ms
		
```

**Comment by ellatrix** [comment]
It's interesting to see how varying the loading times are... almost a 10% faster with this branch, but probably because of some random network or CPU thing?

**Comment by ellatrix** [comment]
Thanks for the reviews everyone! Let's expose this more and iterate if there's any bugs found.

**Comment by jasmussen** [comment]
:keanu-whoah:

ƪ(˘⌣˘)┐ ƪ(˘⌣˘)ʃ ┌(˘⌣˘)ʃ

File: packages/block-editor/src/components/typewriter/index.js
Comment (ellatrix) [reply packages/block-editor/src/components/typewriter/index.js]: 🤯 Didn't realise this could, or is designed to, be used for state as well.

**Comment by ellatrix** [comment]
I'm surprised to see the e2e tests pass locally and on Travis without the need to manually wait the next animation frame after a scroll event. If there turn out to be intermittent failures, it's probably related to the scrolling and can be fixed easily.

**Comment by youknowriad** [comment]
CSS animations are disabled in e2e tests both locally and in travis. Spring-js animations are also disabled on travis and you disable them locally by using an env variable.

**Comment by ellatrix** [comment]
@youknowriad I'm not talking about the CSS animations, but about the animation frame requested during scroll. :)

**Comment by youknowriad** [comment]
😅

**Comment by mapk** [comment]
Beautiful work, @ellatrix! 👏

**Comment by SchneiderSam** [comment]
I just tested the RC-1 and it's a really cool feature. Thanks ;-)

**Pull request merged**
